DAOS-13934 test: Support passing launch.py args per stage

phender opened the pull request:
Provide additional flexibility for passing launch.py tag, provider, and nvme argument values on a per stage and per branch basis.  Existing branch Jenkinsfile implementations will use the existing behavior.

JohnMalmberg commented on vars/getFunctionalArgs.groovy:
```diff
@@ -0,0 +1,49 @@
+ * @param provider default launch.py --provider argument to use
+ * @return Map values to use with runTestFunctional
+ */
+Map call(String pragma_suffix, String stage_tags, String default_tags, String nvme,
```
Generally we use "Map call(Map config = [:])"  If one of these call signatures does not exist, and you are missing one of your parameters, the Jenkins job will crash with a stack dump claiming this routine does not exist.

Then we do the checks on the incoming config Map members to determine if we have all the needed parameters, etc, and report any problems or conflicts.

This is mainly needed for methods that may be called from Jenkinsfiles, as that is more likely where the programming error may show up.

phender replied:
I understand the 'routine does not exist' error - since in grovvy you can define the same function with different inputs, so technically the version w/o all the required inputs does not exist - but using a Map and then having to verify that all the inputs have been specified seems like a lot of extra work just to avoid a failure introduced by a misuse of the function.

phender replied:
The updated implementation now uses Map inputs.

JohnMalmberg commented on vars/getFunctionalArgs.groovy:
```diff
@@ -0,0 +1,49 @@
+    result['stage_rpms'] = ''
+
+    // Get the launch.py --nvme argument from either the build parameters or commit pragmas
+    launch_nvme = params.TestNvme ?: cachedCommitPragma(
```
launch_nvme appears to be a global variable shared by other instances and can changed by them.

Untyped variables are global.  Unfortunately we do not know the scope to intentionally use this feature.

phender replied:
Fixed. Thanks.

JohnMalmberg commented on vars/getFunctionalArgs.groovy:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    // Get the launch.py --provider argument from either the build parameters or commit pragmas
+    launch_provider = ftest_arg_provider ?: params.TestProvider ?: cachedCommitPragma(
```
launch_provider appears to be a global variable.

phender replied:
Fixed. Thanks.

JohnMalmberg commented on vars/getFunctionalArgs.groovy:
```diff
@@ -0,0 +1,49 @@
+    }
+    
+    // Determine any additional rpms needed for this stage
+    if (ftest_arg_provider && ftest_arg_provider.contains('ucx')) {
```
Where does ftest_arg_provider get defined?

phender replied:
Cut-n-paste typo. Fixed. Thanks.

JohnMalmberg reviewed (commented):
Looks like you have some unexpected global variables.

daltonbohning reviewed (commented):
FYI - https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/ci-daos-stack-pipeline-lib-PR-388-weekly-2.4-testing/2/pipeline/
```
startup failed:

/data/jenkins/jobs/daos-stack/jobs/daos/branches/ci-daos-s.lp0p7cmlq84b.4-testing/builds/2/libs/01d212f90e889f6b5c078a130da351ec88f278350c2d709bd14b6e39d9db9bc0/vars/getFunctionalTags.groovy: 50: The current scope already contains a variable of the name stage_tags

 @ line 50, column 12.

       String stage_tags = ''

              ^
```

daltonbohning commented on vars/getFunctionalTags.groovy:
```diff
@@ -0,0 +1,69 @@
+ * @param default_tags default launch.py tags to use when no parameter or commit pragma tags exist
+ * @return String of test tags to run in the stage
+ */
+Map call(String pragma_suffix, String stage_tags, String default_tags) {
+    String requested_tags = ''
+    String stage_tags = ''
```
I think it doesn't like `stage_tags` as an argument and as a local variable

phender replied:
Yeah I fixed it, thanks.  I need to look into getting a groovy linter installed locally.

phender commented:
With these changes we will have the option to define unique functional test behavior per branch and per stage. e.g.
```
steps {
    job_step_update(
        functionalTest(
            inst_repos: daosRepos(),
            inst_rpms: functionalPackages(1, next_version, 'client-tests-openmpi'),
            test_tag: getFunctionalTags(pragma_suffix: '-hw-medium', stage_tags: 'hw,medium', default_tags: 'pr'),
            ftest_arg: getFunctionalArgs(pragma_suffix: '-hw-medium', default_nvme: 'auto', provider: 'ofi+verbs;ofi_rxm'),
            test_function: 'runTestFunctionalV2'))
}
```

phender commented on vars/getFunctionalTags.groovy:
```diff
@@ -0,0 +1,59 @@
+ *      default_tags    launch.py tag argument to use when no parameter or commit pragma tags exist
+ * @return String of test tags to run in the stage
+ */
+Map call(Map kwargs = [:]) {
```
I wonder if I should add optional arguments taht would allow more control over which tags would be used under each started by case: user, upstream, and timer.

phender commented:
Verifying that a PR runs with the correct default tags with these changes in https://github.com/daos-stack/daos/pull/12659

phender commented:
Ticket for the failed release/2.2 downstream PRs: https://daosio.atlassian.net/browse/DAOS-14154

phender commented:
Updated skipStage unit tests are passing (including new EL 9 tests and fixed skip RPM build tests): https://[build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/pipeline-lib/branches/PR-388/runs/19/nodes/52/log/?start=0](https://build.hpdd.intel.com/blue/rest/organizations/jenkins/pipelines/daos-stack/pipelines/pipeline-lib/branches/PR-388/runs/19/nodes/52/log/?start=0)

phender commented:
Failures in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fpipeline-lib/detail/PR-388/20/pipeline/1666/ due to https://daosio.atlassian.net/browse/DAOS-14154

daltonbohning commented on vars/parseStageInfo.groovy:
```diff
@@ -225,140 +189,71 @@ Map call(Map config = [:]) {
-            result['test_tag'] += group + ',' + cluster_size + ' '
-        }
-        result['test_tag'] = result['test_tag'].trim()
+        // Get the ftest arguments
+        kwargs['default_nvme'] = ftest_arg_nvme
+        kwargs['provider'] = ftest_arg_provider
+        functional_args = getFunctionalArgs(kwargs)
+        result['ftest_arg'] = functional_args.get('ftest_arg', '')
+        result['stage_rpms'] = functional_args.get('stage_rpms', '')
```
What's the difference between `kwargs['default_nvme'] = ftest_arg_nvme` and the nvme from  `getFunctionalArgs`?

phender replied:
The `kwargs['default_nvme']` is the `default_nvme` argument passed to `vars/getFunctionalNvme.groovy`.  The `default_nvme` value is used in the return value by `vars/getFunctionalNvme.groovy` when neither the `TestNvme` build parameter nor a `Test-nvme*` commit pragma is set.  Assuming neither build or commit pragma is set, `vars/getFunctionalNvme.groovy` will return `''` for VM stages (because `default_nvme=''`) or `--nvme=auto`, `--nvme=auto_md_on_ssd`, or `--nvme=auto:-3DNAND` for the HW stages (based on the output from `get_default_nvme()`).
